File: packages/sf-game-data/extract/sf-game-data-parse/Recipes.cs
using System.Text.Json;

namespace SfGameData.Parse;

// Lookups built from already-parsed items, resources and buildings.
public sealed class RecipeLookups
{
    public Dictionary<string, string> ItemForm { get; init; } = [];
    public Dictionary<string, string> ItemDisplay { get; init; } = [];
    public Dictionary<string, string> BuildingDisplay { get; init; } = [];
    public HashSet<string> BuildingClasses { get; init; } = [];
}

// A build-gun recipe resolved to the building it constructs and its cost.
public sealed class BuildRecipe
{
    public string BuildingClassName { get; init; } = "";
    public List<BuildCostLine> Cost { get; init; } = [];
}

// Ported from src/parser/extractors/recipes.ts.
public static class Recipes
{
    public static bool IsBuildGunRecipe(JsonElement raw)
        => Util.GetString(raw, "mProducedIn").Contains("BuildGun", StringComparison.OrdinalIgnoreCase);

    private static bool IsAlternateRecipe(string className, string displayName)
        => className.Contains("Alternate", StringComparison.OrdinalIgnoreCase)
            || displayName.StartsWith("Alternate", StringComparison.OrdinalIgnoreCase);

    private static string ResolveForm(RecipeLookups lookups, string className)
        => lookups.ItemForm.GetValueOrDefault(className, "solid");

    private static string ResolveDisplay(RecipeLookups lookups, string className)
        => lookups.ItemDisplay.GetValueOrDefault(className, "");

    private static Ingredient MapIngredient(RawItemAmount rawItem, double craftTime, RecipeLookups lookups)
    {
        var form = ResolveForm(lookups, rawItem.ClassName);
        var display = Normalise.ToDisplayAmount(rawItem.Amount, form);
        return new Ingredient
        {
            ItemClassName = rawItem.ClassName,
            DisplayName = ResolveDisplay(lookups, rawItem.ClassName),
            Amount = display.Amount,
            PerMinute = Normalise.PerMinute(display.Amount, craftTime),
            Unit = display.Unit,
        };
    }

    private static VariablePower? ResolveVariablePower(JsonElement raw)
    {
        var constant = Util.GetNumber(raw, "mVariablePowerConsumptionConstant", 0);
        var factor = Util.GetNumber(raw, "mVariablePowerConsumptionFactor", 1);
        if (constant == 0 && factor == 1)
        {
            return null;
        }
        return new VariablePower { Min = constant, Max = constant + factor };
    }

    /// <summary>Extracts a production recipe. Call only when IsBuildGunRecipe is false.</summary>
    public static Recipe ExtractRecipe(JsonElement raw, RecipeLookups lookups)
    {
        var className = Util.GetString(raw, "ClassName");
        var displayName = Util.GetString(raw, "mDisplayName");
        var craftTime = Util.GetNumber(raw, "mManufactoringDuration", 0);
        var producedInRaw = Util.GetString(raw, "mProducedIn");

        var producedInClasses = ClassRef.ExtractClassNames(producedInRaw)
            .Where(lookups.BuildingClasses.Contains)
            .ToList();
        var producedIn = producedInClasses
            .Select(cn => lookups.BuildingDisplay.GetValueOrDefault(cn, cn))
            .ToList();

        var recipe = new Recipe
        {
            ClassName = className,
            DisplayName = displayName,
            IsAlternate = IsAlternateRecipe(className, displayName),
            CraftTime = craftTime,
            Ingredients = Normalise.ParseItemAmountList(Util.GetString(raw, "mIngredients"))
                .Select(item => MapIngredient(item, craftTime, lookups)).ToList(),
            Products = Normalise.ParseItemAmountList(Util.GetString(raw, "mProduct"))
                .Select(item => MapIngredient(item, craftTime, lookups)).ToList(),
            ProducedIn = producedIn,
            ProducedInClasses = producedInClasses,
            InBuildGun = false,
            InWorkshop = producedInRaw.Contains("WorkBench", StringComparison.OrdinalIgnoreCase)
                || producedInRaw.Contains("Workshop", StringComparison.OrdinalIgnoreCase),
        };

        recipe.VariablePower = ResolveVariablePower(raw);
        return recipe;
    }

    /// <summary>
    /// Resolves a build-gun recipe to the building it constructs via the name
    /// heuristic: product Desc_X_C → building Build_X_C. Returns null when empty.
    /// </summary>
    public static BuildRecipe? ExtractBuildRecipe(JsonElement raw)
    {
        var products = Normalise.ParseItemAmountList(Util.GetString(raw, "mProduct"));
        if (products.Count == 0)
        {
            return null;
        }
        var product = products[0];
        var buildingClassName = product.ClassName.StartsWith("Desc_", StringComparison.Ordinal)
            ? "Build_" + product.ClassName["Desc_".Length..]
            : product.ClassName;
        var cost = Normalise.ParseItemAmountList(Util.GetString(raw, "mIngredients"))
            .Select(item => new BuildCostLine { ItemClassName = item.ClassName, Amount = item.Amount })
            .ToList();
        return new BuildRecipe { BuildingClassName = buildingClassName, Cost = cost };
    }
}

File: packages/sf-game-data/src/parser/extractors/recipes.ts
import type {
  BuildCostLine,
  Ingredient,
  ItemForm,
  Recipe,
  RawClass,
  VariablePower,
} from '../types.js';
import { getNumber, getString } from '../util.js';
import { extractClassNames } from '@foreman/sf-core';
import { parseItemAmountList, type RawItemAmount } from '../normalise/ingredients.js';
import { perMinute, toDisplayAmount } from '../normalise/fluids.js';

/** Lookups built from already-parsed items, resources and buildings. */
export interface RecipeLookups {
  itemForm: Map<string, ItemForm>;
  itemDisplay: Map<string, string>;
  buildingDisplay: Map<string, string>;
  buildingClasses: Set<string>;
}

/** A build-gun recipe resolved to the building it constructs and its cost. */
export interface BuildRecipe {
  buildingClassName: string;
  cost: BuildCostLine[];
}

export function isBuildGunRecipe(raw: RawClass): boolean {
  return /BuildGun/i.test(getString(raw, 'mProducedIn'));
}

function isAlternateRecipe(className: string, displayName: string): boolean {
  return /Alternate/i.test(className) || /^Alternate/i.test(displayName);
}

function resolveForm(lookups: RecipeLookups, className: string): ItemForm {
  return lookups.itemForm.get(className) ?? 'solid';
}

/**
 * The item's authored display name, or `''` when unknown. The edge humanises the
 * bare class name when no authored name exists (presentation boundary).
 */
function resolveDisplay(lookups: RecipeLookups, className: string): string {
  return lookups.itemDisplay.get(className) ?? '';
}

function mapIngredient(
  rawItem: RawItemAmount,
  craftTime: number,
  lookups: RecipeLookups,
): Ingredient {
  const form = resolveForm(lookups, rawItem.className);
  const { amount, unit } = toDisplayAmount(rawItem.amount, form);
  return {
    itemClassName: rawItem.className,
    displayName: resolveDisplay(lookups, rawItem.className),
    amount,
    perMinute: perMinute(amount, craftTime),
    unit,
  };
}

function resolveVariablePower(raw: RawClass): VariablePower | undefined {
  const constant = getNumber(raw, 'mVariablePowerConsumptionConstant', 0);
  const factor = getNumber(raw, 'mVariablePowerConsumptionFactor', 1);
  if (constant === 0 && factor === 1) {
    return undefined;
  }
  return { min: constant, max: constant + factor };
}

/** Extracts a production recipe. Call only when `isBuildGunRecipe` is false. */
export function extractRecipe(raw: RawClass, lookups: RecipeLookups): Recipe {
  const className = getString(raw, 'ClassName');
  const displayName = getString(raw, 'mDisplayName');
  const craftTime = getNumber(raw, 'mManufactoringDuration', 0);
  const producedInRaw = getString(raw, 'mProducedIn');

  const producedInClasses = extractClassNames(producedInRaw).filter((cn) =>
    lookups.buildingClasses.has(cn),
  );
  const producedIn = producedInClasses.map((cn) => lookups.buildingDisplay.get(cn) ?? cn);

  const recipe: Recipe = {
    className,
    displayName,
    isAlternate: isAlternateRecipe(className, displayName),
    craftTime,
    ingredients: parseItemAmountList(getString(raw, 'mIngredients')).map((item) =>
      mapIngredient(item, craftTime, lookups),
    ),
    products: parseItemAmountList(getString(raw, 'mProduct')).map((item) =>
      mapIngredient(item, craftTime, lookups),
    ),
    producedIn,
    producedInClasses,
    inBuildGun: false,
    inWorkshop: /WorkBench|Workshop/i.test(producedInRaw),
  };

  const variablePower = resolveVariablePower(raw);
  if (variablePower !== undefined) {
    recipe.variablePower = variablePower;
  }
  return recipe;
}

/**
 * Resolves a build-gun recipe to the building it constructs. The docs file does
 * not link the build descriptor to its `FGBuildable` entry, so a name heuristic
 * is used: the product `Desc_X_C` maps to the building `Build_X_C`.
 */
export function extractBuildRecipe(raw: RawClass): BuildRecipe | undefined {
  const products = parseItemAmountList(getString(raw, 'mProduct'));
  const product = products[0];
  if (product === undefined) {
    return undefined;
  }
  const buildingClassName = product.className.replace(/^Desc_/, 'Build_');
  const cost = parseItemAmountList(getString(raw, 'mIngredients')).map(
    (item): BuildCostLine => ({ itemClassName: item.className, amount: item.amount }),
  );
  return { buildingClassName, cost };
}
